File: built-in plugins dsp/mb-piano-electric.cpp
/**
 * MB Electric Piano
 * Category : instrument
 * Type     : piano
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Rhodes-style electric piano with bell-like tones
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_PIANO_ELECTRIC_H
#define MB_PIANO_ELECTRIC_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbPianoElectric : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-piano-electric";
    static constexpr const char* PLUGIN_NAME    = "MB Electric Piano";
    static constexpr const char* PLUGIN_TYPE    = "piano";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float bell = 0.5f;  // range [0, 1]
    float tremolo = 0.3f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbPianoElectric() = default;
    ~MbPianoElectric() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.bell = std::clamp(params.bell, 0f, 1f);
        params.tremolo = std::clamp(params.tremolo, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Electric Piano
        return input;
    }
};

#endif // MB_PIANO_ELECTRIC_H

File: built-in plugins dsp/mb-piano-electric.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbPianoElectricPlugin: PluginDefinition = {
    id: 'mb-piano-electric', slug: 'mb-piano-electric', name: 'MB Electric Piano', category: 'instrument', type: 'piano', version: '1.0.0',
    description: 'Rhodes-style electric piano with bell-like tones', author: 'Max Booster',
    oscillators: [{ type: 'sine', detune: 0, gain: 0.7 }, { type: 'triangle', detune: 0.3, gain: 0.3 }],
    envelope: { attack: 0.001, decay: 0.8, sustain: 0.3, release: 0.6 },
    parameters: [
      { id: 'bell', name: 'Bell', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'tremolo', name: 'Tremolo', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { bell: 0.5, tremolo: 0.3, volume: 0.8 },
  };

export default MbPianoElectricPlugin;
